cpp/src/copying/contiguous_split.cu

```diff
@@ -813,282 +945,278 @@ struct size_of_helper {
-  // so we will take the actual set of outgoing source/destination buffers and further partition
-  // them into much smaller chunks in order to drive up the number of blocks and overall occupancy.
-  auto const desired_chunk_size = std::size_t{1 * 1024 * 1024};
-  rmm::device_uvector<thrust::pair<std::size_t, std::size_t>> chunks(num_bufs, stream);
-  thrust::transform(
-    rmm::exec_policy(stream),
-    _d_dst_buf_info,
-    _d_dst_buf_info + num_bufs,
-    chunks.begin(),
-    [desired_chunk_size] __device__(
-      dst_buf_info const& buf) -> thrust::pair<std::size_t, std::size_t> {
-      // Total bytes for this incoming partition
+/**
+ * @brief Get the size in bytes of a batch described by `dst_buf_info`.
+ */
+struct batch_byte_size_function {
+  size_type num_batches;
+  dst_buf_info const* infos;
+  __device__ std::size_t operator()(size_type i) const
+  {
+    if (i == num_batches) {
+      return 0;
+    } else {
+      auto& buf = *(infos + i);
       std::size_t const bytes =
         static_cast<std::size_t>(buf.num_elements) * static_cast<std::size_t>(buf.element_size);
+      return util::round_up_unsafe(bytes, split_align);
+    }
+  }
+};
 
-      // This clause handles nested data types (e.g. list or string) that store no data in the row
-      // columns, only in their children.
-      if (bytes == 0) { return {1, 0}; }
+/**
+ * @brief Get the input buffer index given the output buffer index.
+ */
+struct out_to_in_index_function {
+  offset_type const* batch_offsets;
+  int num_bufs;
+  __device__ int operator()(size_type i) const
+  {
+    return static_cast<size_type>(
+             thrust::upper_bound(thrust::seq, batch_offsets, batch_offsets + num_bufs + 1, i) -
+             batch_offsets) -
+           1;
+  }
+};
 
-      // The number of chunks we want to subdivide this buffer into
-      std::size_t const num_chunks =
-        max(std::size_t{1}, util::round_up_unsafe(bytes, desired_chunk_size) / desired_chunk_size);
+};  // anonymous namespace
 
-      // NOTE: leaving chunk size as a separate parameter for future tuning
-      // possibilities, even though in the current implementation it will be a
-      // constant.
-      return {num_chunks, desired_chunk_size};
-    });
+namespace detail {
 
-  rmm::device_uvector<offset_type> chunk_offsets(num_bufs + 1, stream);
-  auto buf_count_iter = cudf::detail::make_counting_transform_iterator(
-    0, [num_bufs, num_chunks = num_chunks_func{chunks.begin()}] __device__(size_type i) {
-      return i == num_bufs ? 0 : num_chunks(i);
-    });
-  thrust::exclusive_scan(rmm::exec_policy(stream),
-                         buf_count_iter,
-                         buf_count_iter + num_bufs + 1,
-                         chunk_offsets.begin(),
-                         0);
+// packed block of memory 1: split indices and src_buf_info structs
+struct packed_split_indices_and_src_buf_info {
```

Comment:
Does this struct (and it's siblings) have to be in the `detail` namespace because of the exposure in the header?

Reply:
In this case, there is no exposure in the header for these structs. They can be moved to the anonymous namespace, which is what I think the alternative should be (make sure these definitions are private to `contiguous_split.cu`)